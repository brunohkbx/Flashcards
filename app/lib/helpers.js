import { Platform } from 'react-native';

export const waitFor = ms => {
  return new Promise(resolve => setTimeout(resolve, ms));
};

export const getOS = () => {
  return Platform.OS;
};
