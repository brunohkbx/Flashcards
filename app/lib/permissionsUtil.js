import { Permissions } from 'expo';

export const obtainNotificationPermission = async () => {
  const { status: existingStatus } = await Permissions.getAsync(
    Permissions.NOTIFICATIONS
  );
  let finalStatus = existingStatus;

  if (existingStatus !== 'granted') {
    const { status } = await Permissions.askAsync(Permissions.NOTIFICATIONS);

    finalStatus = status;
  }

  return finalStatus;
};
