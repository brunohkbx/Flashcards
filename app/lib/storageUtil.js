import { AsyncStorage } from 'react-native';
import { pickBy, assign } from 'lodash/object';

const DECK_STORAGE_KEY = 'flashcards:deck';

export const getDecks = () => {
  return AsyncStorage.getItem(DECK_STORAGE_KEY);
}

export const createDeck = deck => {
  return AsyncStorage.mergeItem(DECK_STORAGE_KEY, JSON.stringify(deck));
}

export const deleteDeck = id => {
  return getDecks()
    .then(decks => JSON.parse(decks))
    .then(decks => pickBy(decks, deck => deck.id !== id))
    .then(decks => AsyncStorage.setItem(DECK_STORAGE_KEY, JSON.stringify(decks)));
}

export const editDeck = deck => {
  return getDecks()
    .then(decks => JSON.parse(decks))
    .then(decks => assign(decks, deck))
    .then(decks => AsyncStorage.setItem(
      DECK_STORAGE_KEY, JSON.stringify(decks)
    ));
}
