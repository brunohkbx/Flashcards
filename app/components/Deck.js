import React from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  Card,
  CardContent,
  CardActions,
  Colors,
  Title,
  Paragraph
} from 'react-native-paper';
import { capitalize } from '../lib/stringUtil';
import pluralize from 'pluralize';

const Deck = ({ title, flashcardsCount }) => {
  return (
    <Card>
      <CardContent style={{marginHorizontal: 8}}>
        <Title>{capitalize(title)}</Title>
        <Paragraph>{pluralize('card', flashcardsCount, true)}</Paragraph>
      </CardContent>
      <CardActions>
        <Button primary onPress={() => {}}>View</Button>
        <Button color={Colors.pink500} onPress={() => {}}>Delete</Button>
      </CardActions>
    </Card>
  );
}

Deck.propTypes = {
  title: PropTypes.string.isRequired,
  flashcardsCount: PropTypes.number
}

Deck.defaultProps = {
  flashcardsCount: 0
}

export default Deck;
