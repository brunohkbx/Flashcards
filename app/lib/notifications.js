import { Notifications } from 'expo';
import * as storageUtil from '../lib/storageUtil';

export const clearLocalNotification = () => {
  return storageUtil.unscheduleNotification()
    .then(Notifications.cancelAllScheduledNotificationsAsync);
};

export const scheduleLocalNotification = () => {
  return storageUtil.isScheduledNotification()
    .then(JSON.parse)
    .then(data => {
      if (data === null) {
        Notifications.cancelAllScheduledNotificationsAsync();

        let tomorrow = new Date();
        tomorrow.setDate(tomorrow.getDate() + 1);
        tomorrow.setHours(18);
        tomorrow.setMinutes(30);

        Notifications.scheduleLocalNotificationAsync(
          {
            title: 'Haven\'t seen you in a while',
            body: 'ðŸ‘‹ Don\'t forget to take your quiz today!',
            ios: {
              sound: true
            },
            android: {
              sound: true,
              priority: 'high',
              sticky: false,
              vibrate: true
            }
          },
          {
            time: tomorrow,
            repeat: 'day',
          }
        );

        storageUtil.scheduleNotification();
      }
    });
};
