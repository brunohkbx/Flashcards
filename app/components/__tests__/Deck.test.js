import React from 'react';
import { shallow } from 'enzyme';
import Deck from '../Deck';
import pluralize from 'pluralize';
import { capitalize } from '../../lib/stringUtil';

jest.mock('pluralize');
jest.mock('../../lib/stringUtil');

describe('Deck', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      title: 'foo',
      flashcardsCount: 5,
      handleEditPress: jest.fn(),
      handleDeletePress: jest.fn(),
      onPress: jest.fn()
    }, propOverrides)

    const wrapper = shallow(<Deck {...defaultProps} />);

    const editDeck = () => {
      wrapper
        .find('CardActions')
        .childAt(0)
        .simulate('press');
    }

    const deleteDeck = () => {
      wrapper
        .find('CardActions')
        .childAt(1)
        .simulate('press');
    }

    return {
      defaultProps,
      wrapper,
      editDeck,
      deleteDeck
    }
  }

  it('renders properly', () => {
    const { wrapper } =  setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('capitalizes the title', () => {
    const { defaultProps, wrapper } =  setup();

    expect(capitalize).toHaveBeenCalledWith(defaultProps.title);
  });

  it('pluralizes the deck description based on flashcardsCount', () => {
    const { defaultProps, wrapper } =  setup();

    expect(pluralize).toHaveBeenCalledWith('card', defaultProps.flashcardsCount, true);
  });

  test('edit onPress calls handleEditPress', () => {
    const mockHandleEditPress = jest.fn();
    const { editDeck } = setup({ handleEditPress: mockHandleEditPress });

    editDeck();

    expect(mockHandleEditPress).toHaveBeenCalled();
  });

  test('delete onPress calls handleDeletePress', () => {
    const mockHandleDeletePress = jest.fn();
    const { deleteDeck } = setup({ handleDeletePress: mockHandleDeletePress });

    deleteDeck();

    expect(mockHandleDeletePress).toHaveBeenCalled();
  });
});
