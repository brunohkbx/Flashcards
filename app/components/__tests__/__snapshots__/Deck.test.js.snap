// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deck renders properly 1`] = `
<withTheme(Card)>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[MockFunction]}
  >
    <CardContent
      style={
        Object {
          "marginHorizontal": 8,
        }
      }
    >
      <Title />
      <Subheading />
    </CardContent>
    <CardActions>
      <withTheme(Button)
        onPress={[MockFunction]}
        primary={true}
      >
        Edit
      </withTheme(Button)>
      <withTheme(Button)
        onPress={[MockFunction]}
        primary={true}
      >
        Delete
      </withTheme(Button)>
    </CardActions>
  </TouchableOpacity>
</withTheme(Card)>
`;
